### ISSUE TITLE
pytave: 1x1 matrix to scalar fails in lists

### ISSUE THREAD

**Issue by cbm755** [cbm755/octsympy#456]
@genuinelucifer please consider:
````
>> A = python_cmd ('x=Symbol("x"); return Matrix(1,1,[x])')
A = (sym) x
>> A = python_cmd ('x=Symbol("x"); return Matrix(1,1,[x]),')
A = (sym) [x]
>> [A,B] = python_cmd ('x=Symbol("x"); return [Matrix(1,1,[x]), Matrix(1,1,[x])]')
A = (sym) [x]
B = (sym) [x]
````
(those should all be scalars)

**Comment by genuinelucifer** [comment]
Why is scalar needed in the last 2 cases?
I mean if user is storing a list in the 1x1 matrix then he should get the list back!?
I guess a direct change would be to convert all 1 element lists to scalars. Is it desired?

**Comment by cbm755** [comment]
That's how it works right now in master.  

Decision could be revisited.  (But the three cases above should either all return scalar or all return 1x1 matrix.)

**Comment by mtmiller** [comment]
So this is a symbolic var representing a matrix, versus an Octave matrix of symbolic vars, right? I would vote for flattening 1×1 matrices into scalars also, to be consistent with Octave's interpretation.

**Comment by cbm755** [comment]
@mtmiller I'm not sure I follow.  I don't think there is such thing as "an Octave matrix of symbolic vars" (this decision was Issue #1).

I'll summarize:

  * In SymPy, you can have a 1x1 `Matrix` object.

  * In `@double` Octave, you cannot or maybe its more accurate to stay there is no distinction between a 1x1 matrix and a scalar.

  * To be consistent with the `@double` Octave behaviour, `Symbolic` currently automatically converts all 1x1 `Matrix` objects to scalars.  See Issue #15.

  * The current `pytave` branch doesn't *consistently* convert 1x1 `Matrix` to scalar: sometimes it does, sometimes it doesn't.

**Comment by mtmiller** [comment]
Agree, I wasn't aware of the earlier decision, just stating the full possibilities and making sure I understand the landscape. I agree that the `pytave` branch should be consistent with #15.

**Comment by cbm755** [comment]
@genuinelucifer is this still a thing?  I tried:
````
>> b = sym([])
>> b(1) = x
b = (sym) [x]
````
This is causing failures in `symvar`.

**Comment by genuinelucifer** [comment]
I could not figure out the exact cause of it.. I thought it was fixed earlier, but some how resurfaced now.
I went through `check_and_convert` but the loop and function calls appear as they should be.
Probably something in `sym` constructor...
.
Just to be clear.. We want to convert EVERY list/matrix of 1x1 into scalar, right?

**Comment by genuinelucifer** [comment]
So `[[1]]` will become `1` too. and `{1:10; 2: [1] }` will be same as `{1:10; 2: 1 }`

**Comment by cbm755** [comment]
I think the logic is in the old `objectfilter`.  Lists should not be touched.  But `Matrix` objects should.

So no, `[[1]]` (a list of list of `1`) should not be changed.  Its not a matrix.

**Comment by cbm755** [comment]
Here it is in the old python_header code:
````
        if isinstance(x, sp.Matrix) and x.shape == (1, 1):
            return x[0, 0]
````

**Comment by genuinelucifer** [comment]
It seems that after adding the recursion to go through list element, this check was not at its correct place. I moved it and as per the comment [3cd4489](https://github.com/cbm755/octsympy/pull/465/commits/3cd44893ae2d2ad6c7c70324561ea90a4775ab34), this is fixed..

**Comment by cbm755** [comment]
> It seems that after adding the recursion to go through list element,
> this check was not at its correct place. I moved it and as per the

Excellent!  Can you submit a separate PR (to the pytave branch) that 
does this?  Its commit message should have the proper "Fixes Issue #nnn" 
so the issue gets closed automatically.

"git cherry-pick" might be helpful.

**Comment by cbm755** [comment]
FYI:  the "Fixes Issue #nnn" needs to go into the *commit msg* (not into the github comment box).

I.e., 4742475d6633932270660cb5095bb0 could be:
````
Convert all tuples to list to allow element wise edits on the go

Fixes #456
````

**Comment by cbm755** [comment]
(that's just for next time, no action needed)
